Initial functionality for addition paragraph field for Promoted Content on Service Landing Page.

ekes opened the pull request:
Pushing this code to ask some questions, before adding test/upgrade path.

I've included the paragraphs here. Reason: They aren't intended for use anywhere else, paragraphs and dependencies are already required by this module. So it seems to make sense. But I notice that even though I've not got Directories on I have all the paragraphs for Directories. Is the idea to depend on localgov paragraphs and have all paragraphs in use or not in there rather than in the module where they are needed?

I've included in the templates more basic required templates. I'll push the theme ones with the bootstrap classes and link here in a moment. Does this make sense?

Do the field and css classes make sense. If so I can then write some tests. Rather double check first rather than edit them.

Should there be an upgrade path already. At the moment this will install the field new, but won't add it to existing sites. To work like in the future when there are live sites that will update I assume we'll add the field, but disabled.

ekes committed:
Initial functionality for addition paragraph field for Promoted Content on Service Landing Page.

finnlewis commented:
Thanks @ekes 

The new paragraph makes sense to me. Keep it here, keep it clean. But happy to discuss.

I think the naming of the fields makes sense.

The CSS classes also look sensible, but I'm not sure if there are any class naming conventions that I am not aware of (@cjstevens78 @andybroomfield might be able to comment on that)

Upgrade path: 
 - I think this might be a good opportunity to add the field to existing installations as an update hook, to test the process, but also to demostrate new functionality coming into an existing install (Crodyon? London Councils? ). Happy to hear the other side of the argument there too - if it is a lot of work, there are proably few enough sites using it with devs capable of adding the field manually that we could also go down that route. But a case study in adding a field and doing an update would be awesome, no?

stephen-cox commented:
> I notice that even though I've not got Directories on I have all the paragraphs for Directories. Is the idea to depend on localgov paragraphs and have all paragraphs in use or not in there rather than in the module where they are needed?

Are these the homepage paragraphs? There was a discussion between @Adnan-cds  and myself about where these should live as Adnan had added them to the localgov_paragraphs module and they ended up staying there. See https://github.com/localgovdrupal/localgov_paragraphs/pull/6#issuecomment-682462352.

At a minimum they should be probably be moved to a submodule so they can be turned on if desired. This is something @andybroomfield mentioned as desriable (although I can't find the issue for that at the moment).

stephen-cox commented:
> Should there be an upgrade path already. At the moment this will install the field new, but won't add it to existing sites. To work like in the future when there are live sites that will update I assume we'll add the field, but disabled.

I think we should be looking at an upgrade path to add this sort of thing to all existing sites. It'll be much easier to maintain into the future if we know a site with a specific install version has a basic set of fields, even if there may be extra ones as well. The complications come when people have modified settings that we want to update. For example, it's very likely sites will want to change node display settings from the defaults being shipped.

Adnan-cds commented:
> ... Adnan had added them to the localgov_paragraphs module and they ended up staying there.

That's right.  I prefer all Paragraphs to be part of the localgov_paragraphs module or one of its submodules depending on their utility.  That way, anyone who downloads the localgov_paragraphs module should have access to all our Paragraphs.  I find these Paragraphs too useful to be spread out over various other major modules (e.g. localgov_campaigns).  That's what I was discussing with Steve earlier but we didn't eventually do anything about it.

The homepage paragraphs (which have nothing **specific** about homepages really) are part of the localgov_homepage_paragraphs submodule of the localgov_paragraphs module.

> But I notice that even though I've not got Directories on I have all the paragraphs for Directories.

Sorry Ekes, I didn't get it.  I wasn't aware we have Directory specific Paragraphs.  If you please clarify.

andybroomfield commented:
Still doing catch up, but my understanding is like @Adnan-cds says that we don't have directory specific paragraphs (and dont' have them in BHCC at present). I thought the plan was to make all paragraphs into localgov_paragraphs?

ekes committed:
Use core default image style. Add dependencies.

Expectation is that users will change to fit own layout. This will
install without dependencies.

ekes commented:
@finnlewis Updated the dependencies, I've installed it in a clean site from scratch and it went in.

ekes commented:
> we don't have directory specific paragraphs 

Sorry my mistake getting confused.

>  I prefer all Paragraphs to be part of the localgov_paragraphs module

But regarding paragraph configuration:-
We don't have a module for collecting node bundle configurations together. Why do it with paragraph bundles? 
It makes sense to collect together ones that don't have a specific use, and are intended to be used by fields over multiple different entities. 
But if it's a specific paragraph bundle, for a specific field (that config is in the module too, not a fields module), attached a specific entity, why put it into another module?

The precedent is already set I now notice. The other non-reusable field specific paragraph bundle is also in Services for the sublanding page; and that's in the sublanding page module. So this would follow the same pattern.
 
This not only collects the configuration specific to the entity and it's fields together, it makes the git commit much cleaner, atomic. It'd be one squash merge into one module, rather than two issues, two commits, for one field for one bit of functionality.

I would say though that the description of this paragraph, and the one from services sublanding page (which are shown on the paragraphs admin page), and any other specific use paragraph, should clearly state where they are intended to be used (and hence where they come from).

Adnan-cds commented:
> if it's a specific paragraph bundle, for a specific field (that config is in the module too, not a fields module), attached a specific entity, why put it into another module?

I haven't checked the code for this pull request.  Sorry, this is a difficult week.  If you are highly confident that this paragraph has no use outside the context of this module then yes, I am happy to see it being part of this module.  To clarify, for Paragraphs with general use (e.g CTA, Table, etc.), I would prefer them to be part of the localgov_paragraphs module or one of its submodules.

stephen-cox commented:
@ekes Just reviewing merge requests that I've asked to look at. Is there anything here you would like me to look at?

finnlewis commented:
@ekes - remember this? Just wondering if we want to change the base branch to 2.x and take another look?

ekes commented:
> @ekes - remember this? Just wondering if we want to change the base branch to 2.x and take another look?

If there's no product folk out there that are pushing for promoted content it can stay here. If people actually want the feature it looks straightforward enough - although adds another layer of admin.

I'll bump it and keep it open for now?

andybroomfield commented:
I'm not sure what this was, or what it was reffering to. Maybe at this time it's a good place to call it a day with this and close this off, as it would require a lot to resolve the conflicts.

There may be some merit to unbundling the paragraphs that are not shared from localgov_paragraphs, and for having a paragraph field on the landing node (and all pages) but that is better raised as new issues and PRs.

ekes commented:
Happy to close. I can't remember it either, it looks like it was asking questions; and theme at least has moved on a long way.
